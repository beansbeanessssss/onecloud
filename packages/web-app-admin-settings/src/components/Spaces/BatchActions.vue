<template>
  <div>
    <oc-list
      id="oc-spaces-appbar-batch-actions"
      :class="{ 'oc-spaces-appbar-batch-actions-squashed': limitedScreenSpace }"
    >
      <action-menu-item
        v-for="(action, index) in actions"
        :key="`action-${index}`"
        :action="action"
        :items="selectedSpaces"
        appearance="outline"
        class="oc-mr-s"
        :show-tooltip="limitedScreenSpace"
      />
    </oc-list>
  </div>
</template>

<script lang="ts">
import ActionMenuItem from 'web-pkg/src/components/ContextActions/ActionMenuItem.vue'
import Delete from 'web-pkg/src/mixins/spaces/delete'
import Disable from 'web-pkg/src/mixins/spaces/disable'
import Restore from 'web-pkg/src/mixins/spaces/restore'
import { computed, defineComponent, getCurrentInstance, PropType } from 'vue'
import { SpaceResource } from 'web-client'

export default defineComponent({
  name: 'ActionsPanel',
  components: { ActionMenuItem },
  mixins: [Delete, Disable, Restore],
  props: {
    selectedSpaces: {
      type: Array as PropType<Array<SpaceResource>>,
      required: true
    },
    limitedScreenSpace: {
      type: Boolean,
      default: false,
      required: false
    }
  },
  setup(props) {
    const instance = getCurrentInstance().proxy as any
    const actions = computed(() => {
      return [
        ...instance.$_restore_items,
        ...instance.$_delete_items,
        ...instance.$_disable_items
      ].filter((item) => item.isEnabled({ resources: props.selectedSpaces }))
    })
    return { actions }
  }
})
</script>

<style lang="scss">
#oc-spaces-appbar-batch-actions {
  display: block;
  li {
    float: left !important;
  }

  @media only screen and (min-width: 1200px) {
    li {
      margin-top: 0;
      margin-bottom: 0;
    }

    align-items: center;
    display: flex;
  }
}
.oc-spaces-appbar-batch-actions-squashed .oc-files-context-action-label {
  display: none;
}
</style>
