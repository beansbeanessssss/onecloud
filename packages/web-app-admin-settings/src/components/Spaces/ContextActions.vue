<template>
  <div>
    <context-action-menu :menu-sections="menuSections" :items="items" />
    <quota-modal
      v-if="quotaModalIsOpen"
      :cancel="closeQuotaModal"
      :space="quotaModalSelectedSpace"
      @spaceQuotaUpdated="spaceQuotaUpdated"
    />
  </div>
</template>

<script lang="ts">
import Delete from 'web-pkg/src/mixins/spaces/delete'
import Disable from 'web-pkg/src/mixins/spaces/disable'
import EditDescription from 'web-pkg/src/mixins/spaces/editDescription'
import EditQuota from 'web-pkg/src/mixins/spaces/editQuota'
import Rename from 'web-pkg/src/mixins/spaces/rename'
import Restore from 'web-pkg/src/mixins/spaces/restore'
import { computed, defineComponent, getCurrentInstance, PropType, unref } from 'vue'
import { Resource } from 'web-client'
import ContextActionMenu from 'web-pkg/src/components/ContextActions/ContextActionMenu.vue'
import QuotaModal from 'web-pkg/src/components/Spaces/QuotaModal.vue'

export default defineComponent({
  name: 'ContextActions',
  components: { ContextActionMenu, QuotaModal },
  mixins: [Delete, Disable, EditDescription, EditQuota, Rename, Restore],

  props: {
    items: {
      type: Array as PropType<Resource[]>,
      required: true
    }
  },

  setup(props) {
    const instance = getCurrentInstance().proxy

    const filterParams = computed(() => ({ resources: props.items }))
    const menuItemsPrimaryActions = computed(() =>
      [...instance.$_rename_items, ...instance.$_editDescription_items].filter((item) =>
        item.isEnabled(unref(filterParams))
      )
    )
    const menuItemsSecondaryActions = computed(() =>
      [
        ...instance.$_editQuota_items,
        ...instance.$_disable_items,
        ...instance.$_restore_items,
        ...instance.$_delete_items
      ].filter((item) => item.isEnabled(unref(filterParams)))
    )
    const menuSections = computed(() => {
      const sections = []

      if (unref(menuItemsPrimaryActions).length) {
        sections.push({
          name: 'primaryActions',
          items: unref(menuItemsPrimaryActions)
        })
      }
      if (unref(menuItemsSecondaryActions).length) {
        sections.push({
          name: 'secondaryActions',
          items: unref(menuItemsSecondaryActions)
        })
      }

      return sections
    })

    const quotaModalSelectedSpace = computed(() => instance.$data.$_editQuota_selectedSpace)
    const quotaModalIsOpen = computed(() => instance.$data.$_editQuota_modalOpen)
    const closeQuotaModal = () => {
      instance.$_editQuota_closeModal()
    }
    const spaceQuotaUpdated = (quota) => {
      instance.$data.$_editQuota_selectedSpace.spaceQuota = quota
    }

    return {
      menuSections,
      quotaModalSelectedSpace,
      quotaModalIsOpen,
      closeQuotaModal,
      spaceQuotaUpdated
    }
  }
})
</script>
