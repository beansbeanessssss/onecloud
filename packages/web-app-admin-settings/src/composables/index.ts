export * from './actions'
export * from './usePagination'
