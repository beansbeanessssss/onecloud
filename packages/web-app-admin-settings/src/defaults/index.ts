export const supportedLogoMimeTypes = ['image/gif', 'image/jpeg', 'image/png']
export const perPageDefault = '50'
export const perPageStoragePrefix = 'admin-settings'
export const paginationOptions = ['20', '50', '100', '250']
