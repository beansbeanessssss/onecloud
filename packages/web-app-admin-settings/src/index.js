import translations from '../l10n/translations'
import Users from './views/Users.vue'
import Groups from './views/Groups.vue'
import Spaces from './views/Spaces.vue'
// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Administration Settings'),
  id: 'admin-settings',
  icon: 'settings-4',
  isFileEditor: false
}

const routes = [
  {
    path: '/',
    redirect: () => {
      const permissionManager = window.Vue.$permissionManager
      if (permissionManager.hasUserManagement()) {
        return { name: 'admin-settings-users' }
      }
      return { name: 'admin-settings-spaces' }
    }
  },
  {
    path: '/users',
    name: 'admin-settings-users',
    component: Users,
    meta: {
      authContext: 'user',
      title: $gettext('Users')
    }
  },
  {
    path: '/groups',
    name: 'admin-settings-groups',
    component: Groups,
    meta: {
      authContext: 'user',
      title: $gettext('Groups')
    }
  },
  {
    path: '/spaces',
    name: 'admin-settings-spaces',
    component: Spaces,
    meta: {
      authContext: 'user',
      title: $gettext('Spaces')
    }
  }
]

const navItems = [
  {
    name: $gettext('Users'),
    icon: 'user',
    route: {
      path: `/${appInfo.id}/users?`
    },
    enabled: () => {
      const permissionManager = window.Vue.$permissionManager
      return permissionManager.hasUserManagement()
    }
  },
  {
    name: $gettext('Groups'),
    icon: 'group-2',
    route: {
      path: `/${appInfo.id}/groups?`
    },
    enabled: () => {
      const permissionManager = window.Vue.$permissionManager
      return permissionManager.hasUserManagement()
    }
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    route: {
      path: `/${appInfo.id}/spaces?`
    },
    enabled: () => {
      const permissionManager = window.Vue.$permissionManager
      return permissionManager.hasSpaceManagement()
    }
  }
]

export default {
  appInfo,
  routes,
  translations,
  navItems
}
