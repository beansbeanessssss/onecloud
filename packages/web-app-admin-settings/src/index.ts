import translations from '../l10n/translations.json'
import General from './views/General.vue'
import Users from './views/Users.vue'
import Groups from './views/Groups.vue'
import Spaces from './views/Spaces.vue'
// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Administration Settings'),
  id: 'admin-settings',
  icon: 'settings-4',
  isFileEditor: false
}

// FIXME: a better way to access this is needed
const permissionManager = () => (window as any).__$permissionManager

const routes = [
  {
    path: '/',
    redirect: () => {
      if (permissionManager().hasSystemManagement()) {
        return { name: 'admin-settings-general' }
      }
      return { name: 'admin-settings-spaces' }
    }
  },
  {
    path: '/general',
    name: 'admin-settings-general',
    component: General,
    meta: {
      authContext: 'user',
      title: $gettext('General')
    }
  },
  {
    path: '/users',
    name: 'admin-settings-users',
    component: Users,
    meta: {
      authContext: 'user',
      title: $gettext('Users')
    }
  },
  {
    path: '/groups',
    name: 'admin-settings-groups',
    component: Groups,
    meta: {
      authContext: 'user',
      title: $gettext('Groups')
    }
  },
  {
    path: '/spaces',
    name: 'admin-settings-spaces',
    component: Spaces,
    meta: {
      authContext: 'user',
      title: $gettext('Spaces')
    }
  }
]

const navItems = [
  {
    name: $gettext('General'),
    icon: 'settings-4',
    route: {
      path: `/${appInfo.id}/general?`
    },
    enabled: () => {
      return permissionManager().hasSystemManagement()
    }
  },
  {
    name: $gettext('Users'),
    icon: 'user',
    route: {
      path: `/${appInfo.id}/users?`
    },
    enabled: () => {
      return permissionManager().hasUserManagement()
    }
  },
  {
    name: $gettext('Groups'),
    icon: 'group-2',
    route: {
      path: `/${appInfo.id}/groups?`
    },
    enabled: () => {
      return permissionManager().hasUserManagement()
    }
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    route: {
      path: `/${appInfo.id}/spaces?`
    },
    enabled: () => {
      return permissionManager().hasSpaceManagement()
    }
  }
]

export default {
  appInfo,
  routes,
  translations,
  navItems
}
