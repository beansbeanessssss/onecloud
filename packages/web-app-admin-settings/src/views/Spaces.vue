<template>
  <div>
    <app-template
      ref="template"
      :loading="loadResourcesTask.isRunning || !loadResourcesTask.last"
      :breadcrumbs="breadcrumbs"
      :side-bar-active-panel="sideBarActivePanel"
      :side-bar-available-panels="sideBarAvailablePanels"
      :side-bar-open="sideBarOpen"
      @closeSideBar="closeSideBar"
      @toggleSideBar="toggleSideBar"
      @selectPanel="selectPanel"
    >
      <template #topbarActions>
        <div class="admin-settings-app-bar-actions oc-mt-xs">
          <div v-if="selectedSpaces.length > 1" class="oc-flex oc-flex-middle">
            <oc-button
              id="files-clear-selection"
              v-oc-tooltip="$gettext('Clear selection')"
              :aria-label="$gettext('Clear selection')"
              appearance="outline"
              @click="unselectAllSpaces"
            >
              <oc-icon name="close" />
            </oc-button>
          </div>
        </div>
      </template>
      <template #mainContent>
        <no-content-message
          v-if="!spaces.length"
          id="admin-settings-spaces-empty"
          class="spaces-empty"
          icon="layout-grid"
        >
          <template #message>
            <span v-translate>No spaces in here</span>
          </template>
        </no-content-message>
        <div v-else>
          <SpacesList
            :spaces="spaces"
            :selected-spaces="selectedSpaces"
            :header-position="listHeaderPosition"
            @toggleSelectSpace="toggleSelectSpace"
            @toggleSelectAllSpaces="toggleSelectAllSpaces"
            @toggleUnSelectAllSpaces="unselectAllSpaces"
          >
            <template #contextMenu>
              <context-actions :items="selectedSpaces" />
            </template>
          </SpacesList>
        </div>
      </template>
    </app-template>
  </div>
</template>

<script lang="ts">
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import { useAccessToken, useGraphClient, useStore, useTranslations } from 'web-pkg/src/composables'
import { computed, defineComponent, onBeforeUnmount, onMounted, ref, unref } from 'vue'
import { useTask } from 'vue-concurrency'
import { eventBus } from 'web-pkg/src/services/eventBus'
import AppTemplate from '../components/AppTemplate.vue'
import { buildSpace } from 'web-client/src/helpers'
import { configurationManager } from 'web-pkg'
import SpacesList from '../components/Spaces/SpacesList.vue'
import SpaceDetails from 'web-pkg/src/components/sideBar/Spaces/Details/SpaceDetails.vue'
import SpaceDetailsMultiple from 'web-pkg/src/components/sideBar/Spaces/Details/SpaceDetailsMultiple.vue'
import SpaceNoSelection from 'web-pkg/src/components/sideBar/Spaces/SpaceNoSelection.vue'
import ContextActions from '../components/Spaces/ContextActions.vue'
import MembersPanel from '../components/Spaces/SideBar/MembersPanel.vue'

export default defineComponent({
  name: 'SpacesView',
  components: {
    SpacesList,
    AppTemplate,
    NoContentMessage,
    ContextActions
  },
  setup() {
    const store = useStore()
    const accessToken = useAccessToken({ store })
    const spaces = ref([])
    const { graphClient } = useGraphClient()
    const { $gettext } = useTranslations()

    const loadResourcesEventToken = ref(null)
    const template = ref(null)
    const listHeaderPosition = ref(0)
    const selectedSpaces = ref([])
    const sideBarOpen = ref(false)
    const selectedPanel = ref(undefined)

    const loadResourcesTask = useTask(function* (signal) {
      const {
        data: { value: drivesResponse }
      } = yield unref(graphClient).drives.listAllDrives('name asc', 'driveType eq project')
      const drives = drivesResponse.map((space) =>
        buildSpace({ ...space, serverUrl: configurationManager.serverUrl })
      )
      spaces.value = drives
    })

    const breadcrumbs = computed(() => [
      { text: $gettext('Administration Settings'), to: { path: '/admin-settings' } },
      {
        text: $gettext('Spaces'),
        onClick: () => eventBus.publish('app.admin-settings.list.load')
      }
    ])

    const allSpacesSelected = computed(() => unref(spaces).length === unref(selectedSpaces).length)

    const calculateListHeaderPosition = () => {
      listHeaderPosition.value = unref(template)?.$refs?.appBar?.getBoundingClientRect()?.height
    }
    const toggleSelectAllSpaces = () => {
      if (unref(allSpacesSelected)) {
        selectedSpaces.value = []
        return
      }
      selectedSpaces.value = [...unref(spaces)]
    }

    const toggleSelectSpace = (toggledSpace) => {
      const isSpaceSelected = unref(selectedSpaces).find((s) => s.id === toggledSpace.id)
      if (!isSpaceSelected) {
        return selectedSpaces.value.push(toggledSpace)
      }

      selectedSpaces.value = unref(selectedSpaces).filter((s) => s.id !== toggledSpace.id)
    }

    const unselectAllSpaces = () => {
      selectedSpaces.value = []
    }

    const sideBarAvailablePanels = computed(() => {
      return [
        {
          app: 'SpaceNoSelection',
          icon: 'layout-grid',
          title: $gettext('Space details'),
          component: SpaceNoSelection,
          default: true,
          enabled: unref(selectedSpaces).length === 0
        },
        {
          app: 'SpaceDetails',
          icon: 'layout-grid',
          title: $gettext('Space details'),
          component: SpaceDetails,
          default: true,
          enabled: unref(selectedSpaces).length === 1,
          componentAttrs: {
            spaceResource: unref(selectedSpaces)[0]
          }
        },
        {
          app: 'SpaceDetailsMultiple',
          icon: 'layout-grid',
          title: $gettext('Space details'),
          component: SpaceDetailsMultiple,
          default: true,
          enabled: unref(selectedSpaces).length > 1
        },
        {
          app: 'SpaceMembers',
          icon: 'group',
          title: $gettext('Space members'),
          component: MembersPanel,
          default: false,
          enabled: unref(selectedSpaces).length === 1,
          componentAttrs: {
            spaceResource: unref(selectedSpaces)[0]
          }
        }
      ]
    })

    const sideBarActivePanel = computed(() => {
      if (unref(selectedPanel)) {
        return unref(selectedPanel)
      }
      return unref(sideBarAvailablePanels).find((e) => e.enabled).app
    })

    const closeSideBar = () => {
      sideBarOpen.value = false
      selectedPanel.value = false
    }
    const toggleSideBar = () => {
      sideBarOpen.value = !unref(sideBarOpen)
    }
    const selectPanel = (panel) => {
      selectedPanel.value = panel
    }

    onMounted(async () => {
      await loadResourcesTask.perform()
      loadResourcesEventToken.value = eventBus.subscribe('app.admin-settings.list.load', () => {
        loadResourcesTask.perform()
      })

      calculateListHeaderPosition()
      window.addEventListener('resize', calculateListHeaderPosition)
    })

    onBeforeUnmount(() => {
      eventBus.unsubscribe('app.admin-settings.list.load', unref(loadResourcesEventToken))
    })

    return {
      spaces,
      loadResourcesTask,
      graphClient,
      accessToken,
      breadcrumbs,
      listHeaderPosition,
      selectedSpaces,
      sideBarOpen,
      sideBarAvailablePanels,
      sideBarActivePanel,
      closeSideBar,
      toggleSideBar,
      template,
      selectPanel,
      toggleSelectAllSpaces,
      toggleSelectSpace,
      unselectAllSpaces
    }
  }
})
</script>
