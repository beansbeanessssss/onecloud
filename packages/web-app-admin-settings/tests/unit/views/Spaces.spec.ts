import { mockAxiosResolve } from 'web-test-helpers/src/mocks'
import { Graph } from 'web-client'
import { mockDeep } from 'jest-mock-extended'
import { ClientService } from 'web-pkg/src'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  defaultStoreMockOptions,
  mount
} from 'web-test-helpers'
import Spaces from '../../../src/views/Spaces.vue'

const selectors = {
  loadingSpinnerStub: 'apploadingspinner-stub',
  spacesListStub: 'spaceslist-stub',
  noContentMessageStub: 'nocontentmessage-stub'
}

describe('Spaces view', () => {
  describe('loading states', () => {
    it('should show loading spinner if loading', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.find(selectors.loadingSpinnerStub).exists()).toBeTruthy()
    })
    it('should render spaces list after loading has been finished', async () => {
      const { wrapper } = getWrapper()
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.find(selectors.spacesListStub).exists()).toBeTruthy()
    })
  })
  it('should render no content message if no spaces found', async () => {
    const graph = mockDeep<Graph>()
    graph.drives.listAllDrives.mockImplementation(() => mockAxiosResolve({ value: [] }))
    const { wrapper } = getWrapper({ spaces: [] })
    await wrapper.vm.loadResourcesTask.last
    expect(wrapper.find(selectors.noContentMessageStub).exists()).toBeTruthy()
  })
  describe('toggle selection', () => {
    describe('toggleSelectAllSpaces method', () => {
      it('selects all spaces', async () => {
        const spaces = [{ name: 'Some Space' }]
        const { wrapper } = getWrapper({ spaces })
        await wrapper.vm.loadResourcesTask.last
        wrapper.vm.toggleSelectAllSpaces()
        expect(wrapper.vm.selectedSpaces).toEqual(
          expect.arrayContaining([expect.objectContaining({ name: spaces[0].name })])
        )
      })
      it('de-selects all selected spaces', async () => {
        const spaces = [{ name: 'Some Space' }]
        const { wrapper } = getWrapper({ spaces })
        await wrapper.vm.loadResourcesTask.last
        wrapper.vm.selectedSpaces = spaces
        wrapper.vm.toggleSelectAllSpaces()
        expect(wrapper.vm.selectedSpaces.length).toBe(0)
      })
    })
    describe('toggleSelectSpace method', () => {
      it('selects a space', async () => {
        const spaces = [{ name: 'Some Space' }]
        const { wrapper } = getWrapper()
        await wrapper.vm.loadResourcesTask.last
        wrapper.vm.toggleSelectSpace(spaces[0])
        expect(wrapper.vm.selectedSpaces).toEqual(
          expect.arrayContaining([expect.objectContaining({ name: spaces[0].name })])
        )
      })
      it('de-selects a selected space', async () => {
        const spaces = [{ name: 'Some Space' }]
        const { wrapper } = getWrapper()
        await wrapper.vm.loadResourcesTask.last
        wrapper.vm.selectedSpaces = spaces
        wrapper.vm.toggleSelectSpace(spaces[0])
        expect(wrapper.vm.selectedSpaces.length).toBe(0)
      })
    })
    describe('unselectAllSpaces method', () => {
      it('de-selects all selected spaces', async () => {
        const spaces = [{ name: 'Some Space' }]
        const { wrapper } = getWrapper({ spaces })
        await wrapper.vm.loadResourcesTask.last
        wrapper.vm.selectedSpaces = spaces
        wrapper.vm.unselectAllSpaces()
        expect(wrapper.vm.selectedSpaces.length).toBe(0)
      })
    })
  })
})

function getWrapper({ spaces = [{ name: 'Some Space' }] } = {}) {
  const graph = mockDeep<Graph>()
  graph.drives.listAllDrives.mockImplementation(() => mockAxiosResolve({ value: spaces }))
  const $clientService = mockDeep<ClientService>()
  $clientService.graphAuthenticated.mockImplementation(() => graph)
  const mocks = { ...defaultComponentMocks(), $clientService }

  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)

  return {
    wrapper: mount(Spaces, {
      global: {
        plugins: [...defaultPlugins(), store],
        mocks,
        stubs: {
          AppLoadingSpinner: true,
          NoContentMessage: true,
          SpacesList: true
        }
      }
    })
  }
}
