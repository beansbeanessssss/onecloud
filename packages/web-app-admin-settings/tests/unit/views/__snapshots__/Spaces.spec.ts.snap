// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spaces view loading states should render spaces list after loading has been finished 1`] = `
<div>
  <main class="oc-flex oc-flex-wrap oc-height-1-1 app-content oc-width-1-1">
    <div class="oc-width-expand oc-height-1-1 oc-position-relative" id="admin-settings-wrapper">
      <div class="oc-app-bar oc-py-s" id="admin-settings-app-bar">
        <div class="admin-settings-app-bar-controls oc-flex oc-flex-between oc-flex-middle">
          <oc-breadcrumb-stub class="oc-flex oc-flex-middle" contextmenupadding="medium" id="admin-settings-breadcrumb" items="[object Object],[object Object]" maxwidth="-1" showcontextactions="false" truncationoffset="2" variation="default"></oc-breadcrumb-stub>
          <portal-target name="app.runtime.mobile.nav"></portal-target>
          <div class="oc-flex">
            <div class="oc-flex oc-flex-middle">
              <!--v-if-->
              <button aria-label="Display customization options of the files list" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-my-s oc-p-xs" id="files-view-options-btn" type="button">
                <!-- @slot Content of the button -->
                <span class="oc-icon oc-icon-m oc-icon-passive">
                  <!---->
                </span>
              </button>
              <div class="oc-drop oc-box-shadow-medium oc-rounded oc-width-auto" id="files-view-options-drop">
                <div class="oc-card oc-card-body oc-background-secondary oc-p-m">
                  <ul class="oc-list oc-my-rm oc-mx-rm">
                    <!--v-if-->
                    <!--v-if-->
                    <li class="files-view-options-list-item">
                      <div class="oc-page-size files-pagination-size">
                        <label class="oc-page-size-label" for="oc-page-size-3">Items per page</label>
                        <div class="oc-page-size-select" input-id="oc-page-size-3" options="20,50,100,250">
                          <label class="oc-label" for="oc-select-4"></label>
                          <div class="v-select vs--single vs--unsearchable oc-select oc-page-size-select" dir="auto" style="background: transparent;">
                            <div aria-expanded="false" aria-label="Search for option" aria-owns="vs2__listbox" class="vs__dropdown-toggle" id="vs2__combobox" role="combobox">
                              <div class="vs__selected-options">
                                <input aria-autocomplete="list" aria-controls="vs2__listbox" aria-labelledby="vs2__combobox" autocomplete="off" class="vs__search" id="oc-select-4" readonly="" type="search">
                              </div>
                              <div class="vs__actions">
                                <button aria-label="Clear Selected" class="vs__clear" style="display: none;" title="Clear Selected" type="button">
                                  <svg height="10" width="10" xmlns="http://www.w3.org/2000/svg">
                                    <path d="M6.895455 5l2.842897-2.842898c.348864-.348863.348864-.914488 0-1.263636L9.106534.261648c-.348864-.348864-.914489-.348864-1.263636 0L5 3.104545 2.157102.261648c-.348863-.348864-.914488-.348864-1.263636 0L.261648.893466c-.348864.348864-.348864.914489 0 1.263636L3.104545 5 .261648 7.842898c-.348864.348863-.348864.914488 0 1.263636l.631818.631818c.348864.348864.914773.348864 1.263636 0L5 6.895455l2.842898 2.842897c.348863.348864.914772.348864 1.263636 0l.631818-.631818c.348864-.348864.348864-.914489 0-1.263636L6.895455 5z"></path>
                                  </svg>
                                </button>
                                <svg class="vs__open-indicator" height="10" role="presentation" width="14" xmlns="http://www.w3.org/2000/svg">
                                  <path d="M9.211364 7.59931l4.48338-4.867229c.407008-.441854.407008-1.158247 0-1.60046l-.73712-.80023c-.407008-.441854-1.066904-.441854-1.474243 0L7 5.198617 2.51662.33139c-.407008-.441853-1.066904-.441853-1.474243 0l-.737121.80023c-.407008.441854-.407008 1.158248 0 1.600461l4.48338 4.867228L7 10l2.211364-2.40069z"></path>
                                </svg>
                                <div class="vs__spinner" style="display: none;">Loading...</div>
                              </div>
                            </div>
                            <transition-stub appear="false" css="true" name="vs__fade" persisted="false">
                              <ul id="vs2__listbox" role="listbox" style="display: none; visibility: hidden;"></ul>
                            </transition-stub>
                          </div>
                          <!--v-if-->
                        </div>
                      </div>
                    </li>
                    <!--v-if-->
                  </ul>
                </div>
              </div>
            </div>
            <button aria-label="Open sidebar to view details" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-my-s oc-p-xs" id="files-toggle-sidebar" type="button">
              <!-- @slot Content of the button -->
              <span class="oc-icon oc-icon-m oc-icon-passive">
                <!---->
              </span>
            </button>
          </div>
        </div>
        <div class="oc-flex oc-flex-middle oc-mt-xs">
          <div class="admin-settings-app-bar-actions">
            <!--v-if-->
          </div>
          <!--v-if-->
        </div>
      </div>
      <!--v-if-->
      <div>
        <spaces-list-stub class="" selectedspaces="" spaces="[object Object]"></spaces-list-stub>
      </div>
    </div>
    <!--v-if-->
  </main>
</div>
`;
